File: 兴趣班入门百练讲义/NQ100/NQ072/NQ072_improved.py
def is_safe(row, col, positions):
    for i in range(row):
        if positions[i] == col: return False  # 检查列冲突
        if abs(positions[i] - col) == abs(row - i): return False  # 检查对角线冲突
    return True

def dfs(row, N, positions):
    if row == N:
        print(''.join(map(str, positions)))
        return
    for col in range(1, N + 1):
        if is_safe(row, col, positions):
            positions[row] = col
            dfs(row + 1, N, positions)  # 递归搜索下一行

def main():
    N = int(input().strip())
    positions = [0] * N
    dfs(0, N, positions)

if __name__ == "__main__":
    main()

File: source/algorithm/NQ072/NQ072_improved.cpp
#include <iostream>
#include <vector>
using namespace std;
int N;
vector<int> positions;
bool isSafe(int row, int col) {
    for (int i = 0; i < row; ++i) {
        if (positions[i] == col) return false; // 检查列冲突
        if (abs(positions[i] - col) == abs(row - i)) return false; // 检查对角线冲突
    }
    return true;
}
void dfs(int row) {
    if (row == N) {
        for (int pos : positions) cout << pos;
        cout << endl;
        return;
    }
    for (int col = 1; col <= N; ++col) {
        if (isSafe(row, col)) {
            positions[row] = col;
            dfs(row + 1); // 递归搜索下一行
        }
    }
}
int main() {
    cin >> N;
    positions.resize(N);
    dfs(0);
    return 0;
}
